src/models/tabular_flows.jl

```diff
@@ -0,0 +1,115 @@
+using Flux
+using StatsBase
+using Distributions
+using DistributionsAD
+using ContinuousFlows: RealNVP, MaskedAutoregressiveFlow
+using ValueHistories
+using MLDataPattern: RandomBatches
+
+abstract type TabularFlow end
+
+struct RealNVPFlow <: TabularFlow
+	flows
+	base
+end
+
+function RealNVPFlow(nflows::Int, isize::Int, hsize::Int, nlayers::Int)
+	RealNVPFlow(Chain([
+		RealNVP(
+			isize, 
+			(d, o, ftype, postprocess) -> build_mlp(d, hsize, o, nlayers, ftype=ftype, lastlayer=postprocess),
+			mod(i,2) == 0)
+		for i in 1:nflows]...), MvNormal(isize, 1.0f0))
+end
+
+(nvpf::RealNVPFlow)(X) = nvpf.flows(X)
+Flux.trainable(nvpf::RealNVPFlow) = (nvpf.flows, )
+
+struct MAF <: TabularFlow
+	flows
+	base
+end
+
+function MAF(nflows::Int, isize::Int, hsize::Int, nlayers::Int, ordering::String)
+	MAF(Chain([
+        MaskedAutoregressiveFlow(
+            isize, 
+            hsize,
+            nlayers, 
+            isize, 
+            (ordering == "natural") ? (
+                (mod(i, 2) == 0) ? "reversed" : "sequential"
+              ) : "random"
+            ) 
+        for i in 1:nflows]...), MvNormal(isize, 1.0f0))
+end
+
+(maf::MAF)(X) = maf.flows(X)
+Flux.trainable(maf::MAF) = (maf.flows, )
+
+
+function loss(model::F, X) where {F <: TabularFlow}
+    Z, logJ = model((X, _init_logJ(X)))
+    -sum(logpdf(model.base, Z)' .+ logJ)/size(X, 2)
+end
+
+function StatsBase.fit!(model::F, data::Tuple, p) where F <: TabularFlow
+	opt = Flux.ADAM(p.lr)
+	
+	trn_model = deepcopy(model)
+	ps = Flux.params(trn_model);
+
+	X = data[1][1]
+	X_val = data[2][1]
+
+	history = MVHistory()
+	patience = p.patience
+	reg = (p.wreg > 0) ? l2_reg : _ -> 0.0
+	
+	best_val_loss = loss(trn_model, X_val)
+	i = 1
+	for batch in RandomBatches(X, p.batchsize)
```

Comment:
Does this mean that the loop runs indefinitely until convergence criterion is satisfied? What if it never is?

Reply:
I though of giving it some upper limit like 5000, which is slightly above all the models that I have trained so far. Based on my observation 95% of models finish before that condition is met. These models like to overfit, so validation degrades reliably.

With all this said I will add some limit, in order to get the model before slurm kills after 24h, it would be wasteful.

Reply:
The time limit seems to make the most sense.